File: 16.js
/* Given an array nums of distinct integers, return all the possible permutations. You can return the answer in any order.

Example 1:

Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
Example 2:

Input: nums = [0,1]
Output: [[0,1],[1,0]]
Example 3:

Input: nums = [1]
Output: [[1]]

Constraints:

1 <= nums.length <= 6
-10 <= nums[i] <= 10
All the integers of nums are unique. */

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  let solutions = [];

  const recurse = (current, used) => {
    if(current.length === nums.length) {
      solutions.push(current); // if we have a full solution push it to the res array and move on
    } else {
      for(let i = 0; i < nums.length; i++) {
        if(used.indexOf(i) === -1) { // Check if we have used the idex already
          recurse((current + (nums[i])), used + i); // add it to current
        }
      }
    }
  }
  recurse('', '');
  return solutions;
};

console.log(permute([1,2,3]))

File: python/16.py

# var permute = function(nums) {
#   let solutions = [];

#   const recurse = (current, used) => {
#     if(current.length === nums.length) {
#       solutions.push(current); // if we have a full solution push it to the res array and move on
#     } else {
#       for(let i = 0; i < nums.length; i++) {
#         if(used.indexOf(i) === -1) { // Check if we have used the idex already
#           recurse((current + (nums[i])), used + i); // add it to current
#         }
#       }
#     }
#   }
#   recurse('', '');
#   return solutions;
# };

# console.log(permute([1,2,3]))

class permutes(object):
    def permute(nums):
        # for id1, num in enumerate(nums):
        #   for id2, num2 in enumerate(nums):
        #     if num + num2 == target and id1 != id2:
        #       return [id1, id2]
        solutions = []
        def recurse(current, used):
          # print(len(current), len(nums))
          if len(current) == len(nums):
            solutions.append(current)
          else:
            for num in nums:
              num = str(num)
              if used.find(num) == -1:
                recurse((current + num), (used + num))
        recurse('', '')
        return solutions

    print(permute([1,2,3]))